flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -762,17 +762,25 @@ private void doRun() {
             executingThread.setContextClassLoader(userCodeClassLoader.asClassLoader());
 
             AbstractInvokable finalInvokable = invokable;
-            runWithSystemExitMonitoring(finalInvokable::restore);
+            try {
```

Comment:
Can you extract at least some of this code to smaller methods? `doRun()` is already too large, and you are making it even larger. (you don't have to fix all of it (would be great if you want to ;) ), but at least some of it